### ISSUE TITLE
More keycode uses come to light with recent update.

### ISSUE THREAD

**Issue by Brandon-Beck** [exiva/Vizio_SmartCast_API#8]
Here is what I have figured out so far in the last 2-3 days since I found your wonderful API documentation 

DPAD
- 0 down
- 1 left
- 2 select
- 7 right
- 8 up

NAV (guessing here)
- 0 back (probably)
- 3 exit (probably)
- 4 Close Captions toggle
- 6 show status bar overlay
- 8 toggle display setting panel
- 15 Seems to Go to home/launch page, auto switching to CAST input, closes any open streams
- 16 Seems to Go to home/launch page, auto switching to CAST input, closes any open streams
- 17 activate store demo mode. (can cancel during ~10 second confirmation delay. undo it with remote, and turn off in settings in smartcast app. I think 0 or 3 can also cancel it as well)

I got nothing after 17.

TRANSPORT

- 2 toggle play
- 9 closed stream/disconnected. Maybe Stop button?

Everything after 10 for TRANSPORT returned URI NOT FOUND.
not sure what other things between 0-10 are. Perhaps testing again with netflix instead of chromecast will reveal rewind/ff, audio lang, subtitle lang, sd/hq video or something. But I havn't setup the tv for that since the update. I always use chromecast.

On that note, I would assume ASCII will be useful for logging in. though I haven't looked into it yet.

there is also a '/oobe/currentstate', though I have no clue what oobe stands for, may be useless.

Still looking to see if I can pull any more use out of it. Hoping to at find a short 10 second rewind button like the app has on my phone, if that works from a non-casting device. I used this API and parts of pyvizio to write a alexa skill to control the tv. It only works for me since it requires a bridge on the same network as the tv. Probably won't publish it since Vizio plans on releasing an alexa app soon.

p.s. Thanks for documenting this! Our tv remote is lost in one of our 11 couches, and crunchyroll always steps the volume up/down by 7 which was always too loud or quiet. All my old frustrations fade away with
Alexa, ask vizio to set the volume to 20

**Comment by exiva** [comment]
I assume most of these newly used controls are for the new Smartcast functionality?

**Comment by Brandon-Beck** [comment]
yes, for the new smart-tv interface. You can test the D-PAD on the new home screen, which you can get to with nav 15 or 16. The DPAD can also be used for navigating menus like the display settings panel.
I discovered their uses by spamming the TV, you may be able to pull more by dissecting the app source.

Which reminds me, there is also a /menu_native/static which seems to shows you all possible values for T_LIST types. Could be used to dynamically create menus for tv settings, with dropdown options for T_LIST types containing only values that tv supports.
Eg. 
**/menu_native/static/tv_settings/audio/digital_audio_out**
```
    "ITEMS": [
        {
            "TYPE": "T_LIST_V1",
            "CNAME": "digital_audio_out",
            "GROUP": "G_AUDIO",
            "NAME": "Digital Audio Out",
            "ELEMENTS": [
                "Auto",
                "PCM",
                "Dolby D",
                "Bitstream"
            ]
        }
    ]
```
**/menu_native/dynamic/tv_settings/audio/digital_audio_out**
```
    "ITEMS": [
        {
            "HASHVAL": 655175175,
            "CNAME": "digital_audio_out",
            "TYPE": "T_LIST_V1",
            "NAME": "Digital Audio Out",
            "VALUE": "PCM"
        }
    ]
```

**Comment by jkantr** [comment]
Sorry to barge into this thread, I think it's somewhat related?

So when they pushed the new smartcast on-screen stuff, they also sent me a free remote, the [XRT136](https://support.vizio.com/s/article/XRT136-Remote-Information-EXT?language=en_US)

I'm *preeeeety sure* it's not IR... and the remote has one-touch buttons for apps like Netflix and Hulu. Is it possible there are codes for these also? I use this documentation for an echo app, would love to get commands in to launch with voice ^_^

**Comment by Brandon-Beck** [comment]
@jkantr Note that the remote specs you linked state

| Key Features | |
|------|--------|
| IR Front: | Yes| 
| IR Side: | No |
| Bluetooth: | No |
| WiFi Direct: | No |
| Backlit: | No |
| Qwerty Keyboard: | No |

You can test by covering the front of the remote with your hand and pushing buttons, but it probably is not. One touch access to Netflix is common across most vizio smart-tv models with their IR remotes. I have 1 with a keyboard, IR Front, and IR Side (the side IR allows you to hold the remote sideways while typing on the keyboard).

But I believe it is a planned feature that we will be able to exploit soon, if it isn't already implemented. The Launch Codeset seems promising, but I was unable to get anything to happen by spamming it with the normal Integer values of 0-50 when I tested. And there does not yet seem to be a way to launch Netflix from the android app directly, or at least it doesnt allow you to do so without Netflix installed on the phone.

**Comment by jkantr** [comment]
Ah I see, alright this all makes sense, thank you. Sorry I missed it in the specs.

**Comment by Brandon-Beck** [comment]
As a workaround, you could launch netflix by sending Nav 15, and then a few DPAD commands. Unfortunatly, DPAD doesn't seem to play well with keylist array, or rather, it ignores multiple DPAD commands until the TV starts the swipe animation, I think. So you can automate it, but you still have to look at the Home screen for a couple of seconds

**Comment by midN** [comment]
@Brandon-Beck Can you share an curl example on what data are u exactly sending to control DPAD/NAV?

Is it the `key_command` api? What are the codesets for DPAD/NAV then?
